Implemented real-time updates for private chats

Nath-Assa opened the pull request:
### Key changes:

- Implemented `observeMessages(user1: String, user2: String)` in both Mock and Firebase message repositories to enable real-time updates in chat interactions.
-  Developed a private `observeMessages()` function within `ChatViewModel` to encapsulate the logic for observing chat messages. This function is called in the init block.
- Wrote comprehensive tests for all functions to ensure they handle both success and failure scenarios effectively.
- Conducted a live application testing to validate the functionality and real-time performance of the messaging feature.

jgeaso1266 commented on app/src/main/java/com/github/se/eventradar/viewmodel/ChatViewModel.kt:
```diff
@@ -70,6 +70,7 @@ constructor(
         val userId = userRepository.getCurrentUserId()
 
         if (userId is Resource.Success) {
+          observeMessages(userId.data, opponentId)
           it.copy(userId = userId.data)
         } else {
           Log.d(
```
It is unconventional to have all of this logic in the `update` block. To separate the act of updating the UI state and getting the user ID, this logic should be outside of the `update` block. If there is a particular reason that you are doing this please let me know (and leave a comment in the code if applicable)

Nath-Assa replied:
Do you mean init block?

jgeaso1266 reviewed (commented):
One comment. Also, you put live app testing in your TODOs in the description. That should be done before merging.

Nath-Assa commented:
> One comment. Also, you put live app testing in your TODOs in the description. That should be done before merging.

Yes, this is planned before merging.

christinehahaa commented on app/src/test/java/com/github/se/eventradar/ChatViewModelUnitTest.kt:
```diff
@@ -270,4 +272,60 @@ class ChatViewModelUnitTest {
+    delay(500)
+    val expectedLogMessage = "Failed to fetch messages: ${exception.message}"
+
+    val expectedMH =
```
These will likely need to be updated after the message repo update, will communicate offline for help with this

Nath-Assa commented:
**Live Application Testing**:
The listener seems to work perfectly. In the Firestore Database, I added a message to a message history's message list and updated the message history's latest message id. The changes were immediately reflected in the app without any manual intervention. 
![image](https://github.com/swent-sp-2024-event-radar/event-radar-app/assets/126282075/be1e620d-6808-4c98-b596-809c683cf115)
![image](https://github.com/swent-sp-2024-event-radar/event-radar-app/assets/126282075/e8c1d3ee-89c1-4286-a3f4-6954c7507913)
![image](https://github.com/swent-sp-2024-event-radar/event-radar-app/assets/126282075/afce1ab2-882b-4338-a8f5-fdaeb2bf7f8a)

Nath-Assa commented:
Live Application Testing pt2:
@jgeaso1266 and I both ran the app simultaneously and sent each other private messages. The app automatically updated the chats and behaved as expected.

christinehahaa reviewed (approved):
Nice job fixing bugs and testing. CI just needs to pass, then LGTM

Nath-Assa commented:
After the fixes, I have conducted another final live app test and the live updates still work perfectly.
